actors/evm/src/interpreter/system.rs

```diff
@@ -31,30 +41,171 @@ pub enum StorageStatus {
+            None => self.set_bytecode(&[])?,
+        };
+        let new_root = self
+            .rt
+            .store()
+            .put_cbor(
+                &State {
+                    bytecode: bytecode_cid,
+                    contract_state: self.slots.flush().context_code(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        "failed to flush contract state",
+                    )?,
+                    nonce: self.nonce,
+                },
+                Code::Blake2b256,
+            )
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to write contract state")?;
+
+        self.rt.set_state_root(&new_root)?;
+        self.saved_state_root = Some(new_root);
+        Ok(())
+    }
+
+    /// Reload the actor state if changed.
+    pub fn reload(&mut self) -> Result<(), ActorError> {
+        let root = self.rt.get_state_root()?;
+        if self.saved_state_root == Some(root) {
+            return Ok(());
+        }
+        let state: State = self
+            .rt
+            .store()
+            .get_cbor(&root)
+            .context_code(ExitCode::USR_SERIALIZATION, "failed to decode state")?
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "state not in blockstore")?;
+
+        self.slots
+            .set_root(&state.contract_state)
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "state not in blockstore")?;
+        self.nonce = state.nonce;
+        self.saved_state_root = Some(root);
+        self.bytecode = Some(state.bytecode);
+        Ok(())
     }
 
-    pub fn flush_state(&mut self) -> Result<Cid, ActorError> {
-        self.state.flush().map_err(|e| ActorError::illegal_state(e.to_string()))
+    /// Load the bytecode.
+    pub fn load_bytecode(&self) -> Result<Option<Bytecode>, ActorError> {
+        match &self.bytecode {
+            Some(cid) => {
+                let bytecode = self
+                    .rt
+                    .store()
+                    .get(cid)
+                    .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to read state")?
+                    .expect("bytecode not in state tree");
+                if bytecode.is_empty() {
+                    return Ok(None);
+                }
+                Ok(Some(Bytecode::new(bytecode)))
+            }
+            None => Ok(None),
+        }
+    }
+
+    /// Set the bytecode.
+    pub fn set_bytecode(&mut self, bytecode: &[u8]) -> Result<Cid, ActorError> {
+        self.saved_state_root = None;
+        // Reject code starting with 0xEF, EIP-3541
+        if bytecode.first() == Some(&0xEF) {
+            return Err(ActorError::illegal_argument(
+                "EIP-3541: Contract code starting with the 0xEF byte is disallowed.".into(),
+            ));
+        }
+        let k = self
+            .rt
+            .store()
+            .put(Code::Blake2b256, &Block::new(IPLD_RAW, bytecode))
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to write bytecode")?;
```

Comment:
nit: Is this really illegal state if this fails to push to dirty state?

Reply:
Well, it means that the state _would_ be illegal. More importantly, it's the error we tend to return elsewhere.